Fix key shortcuts for quickly switching plot axis

GuiMacielPereira opened the pull request:
Reported by Maximilian Skoda, ISIS

### Description of work
Previous key shortcut behaviour was the default matplotlib shortcuts, which were flakey in a lot of plots.
Fixes #37402 

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
-->
I disabled the default keymaps and implemented our own event for key presses.

#### Purpose of work
Correct scaling of axis for MantidAxes and non-MantidAxes plots by using the keys `k` and `l`.


<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->

### To test:
Run the script in #37402 and check that issue is fixed when switching axes by pressing `k` and `l`.
(For the shortcuts to work, your cursor needs to be on top of the plot)

- Check that shortcuts do nothing for 3D plots
- Open Colorfill and check that if the cursor is on top of the colorbar, nothing happens when you press the shortcuts
- Open a plot with tiled subplots and check that shortcuts work.
- Build documentation and check that table of shortcuts looks okay

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

GuiMacielPereira committed:
Add key shortcuts 'k' and 'l' for axis scale
Disabled the key shortcuts for changing the axis scale in matplotlib by default.
Introduced a key press event so that keys 'k' and 'l' are now handled by us.
The axis to be updated is picked from the position of the cursor.

Made this change because previous default behaviour from matplotlib was unreliable,
and it makes more sense to update axis in the same way we update them through the dialogs.

- If multiple plots, only plot with the cursor on top of it will be updated
- Colorfills work as expected, colorbar does not get updated
- 3D plots are ignored
- Normal mpl plots (non MantidAxis) will have default handling overwritten.

GuiMacielPereira committed:
Get axes images and lines through getter method

Should stick to getter methods if possible.

GuiMacielPereira committed:
Add unit test

Added unit test for checking that a 'k' key press changes the right axis.
Added call() to list of expected calls since I introduced am mlp connection
to a key press event.
The line that adds a "fake_line" is necessary so that if statement inside
on_key_press() doesn't prevent the rest of the function from executing.

GuiMacielPereira committed:
Add release note

GuiMacielPereira committed:
Add documentation table of shortcuts

Checked keymaps table in the default rcParams in matplolib.
Tested keymaps that work on Mantid and wrote a similar table in our documentation.

GuiMacielPereira commented on qt/applications/workbench/workbench/plotting/config.py:
```diff
@@ -39,6 +39,10 @@ def initialize_matplotlib():
+    # Disabling default key shortcuts for toggling axes scale
+    mpl.rcParams["keymap.xscale"].remove("k")
+    mpl.rcParams["keymap.xscale"].remove("L")
+    mpl.rcParams["keymap.yscale"].remove("l")
```
This disables the default keymaps, I am not sure if this is the best place to put it?

jhaigh0 replied:
do you need to remove upper case L?

GuiMacielPereira replied:
I think so, because `L` does the same thing as `k` in the original keymaps. So if I leave it in, `k` will be handled by us but `L` will default to the previous flakey behaviour

jhaigh0 replied:
okay 👍 and just to check `K` doesn't do the same thing as `l`?

GuiMacielPereira replied:
Yeah I just ckecked, `K` doesn't do anything :+1:

jhaigh0 commented on docs/source/release/v6.11.0/Workbench/Bugfixes/37438.rst:
```diff
@@ -0,0 +1 @@
+- Fixed key shortcuts `k` and `l` in mantid plot for quickly switching between `linear` and `log` axis.
```
```suggestion
- Fixed key shortcuts `k` and `l` in mantid plot for quickly switching between `linear` and `log` scales.
```

jhaigh0 reviewed (changes requested):
A few small things. The other commands seem to work well too.

Switching back and forth between log and linear can end up where the plot looses the border we put around it, but this is unrelated to this pr and can be fixed by clicking home (or pressing 'h').

GuiMacielPereira commented on qt/applications/workbench/workbench/plotting/figureinteraction.py:
```diff
@@ -144,6 +147,28 @@ def on_scroll(self, event):
+            next_yscale = self._get_next_axis_scale(current_yscale)
+            self._quick_change_axes((ax.get_xscale(), next_yscale), ax)
+
+    def _get_next_axis_scale(self, current_scale):
+        available_scales = AXES_SCALE_DEFAULT_SWITCH
+        scale_index = available_scales.index(current_scale) if current_scale in available_scales else 0
+        next_scale = available_scales[(scale_index + 1) % len(available_scales)]
+        return next_scale
```
Yeah, I had the suspicion I was overcomplicating it, so you're right I'll change it

GuiMacielPereira committed:
Fix typo in release note

Co-authored-by: Jonathan Haigh <35813666+jhaigh0@users.noreply.github.com>

GuiMacielPereira committed:
Simplify function to change axis

Previous proposal was over-engineered with the idea of
possibly extending the number of scales the axis could scale to.
Since we're likely to just stick to linear and log, the
function is now easier to understand and maintain.

GuiMacielPereira committed:
Add unit test for `l` shortcut

Since I already had a test for `k`, added a similar test for `l`.

jhaigh0 commented:
Meant to save in the approval message that I've checked that this is integrating well with the plot settings and axis edit menus.
